### PR TITLE
[ConstraintSystem] Invalidate solution in presence of error types

### PR DESCRIPTION
**Pull request by xedin**
In `ConstraintSystem::finalize` if simplification of one of the
type variables present in the system results in introduction of
error type to the possible solution, let's not return that solution,
because that might cause problems down the road in verifier.

One of the reasons why `simplifyType` could produce an error type
is related to substitution of associated types using newly
discovered base type because its protocol conformance as not
properly verified until later on.

Resolves [SR-4575](https://bugs.swift.org/browse/SR-4575).

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by slavapestov** [comment]
This fix LGTM.

File: validation-test/compiler_crashers_2_fixed/0097-sr4575.swift [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  func bar<T>(_ to: T.Type) -> V<T> {
+    return V<T>()
+  }
+}
+
+extension S {
+  func foo<R>(_ body: (UnsafeBufferPointer<UTF16.CodeUnit>) -> R) -> R {
+    return Array(self.bar(UTF16.self)).withUnsafeBufferPointer(body)
+  }
+}
```
Comment (rudkx) [inline validation-test/compiler_crashers_2_fixed/0097-sr4575.swift]: newline

**Comment by xedin** [comment]
/cc @DougGregor

@rudkx @slavapestov I've updated the patch so `ConstraintSystem::finalize` is going to produce `Optional<Solution>` which means that if we detect that there is an error type present in one of the bindings we are going to produce no solution.

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test Linux platform

**Comment by rudkx** [comment]
@xedin Pavel, @slavapestov and I were just discussing this and think there's something else fundamentally wrong that should be fixed here with hasTypeParameter(). He's looking at fixing that, which resolve the assert hit in the original test case.

**Comment by xedin** [comment]
@rudkx @slavapestov so should I close this then?

**Pull request closed without merging**